<!DOCTYPE html>

<html>

<head>
    <meta charset="utf-8">
    <!--
        Customize this policy to fit your own app's needs. For more guidance, please refer to the docs:
            https://cordova.apache.org/docs/en/latest/
        Some notes:
            * https://ssl.gstatic.com is required only on Android and is needed for TalkBack to function properly
            * Disables use of inline scripts in order to mitigate risk of XSS vulnerabilities. To change this:
                * Enable inline JS: add 'unsafe-inline' to default-src
                 <meta http-equiv="Content-Security-Policy"
        content="default-src 'self' data: https://ssl.gstatic.com 'unsafe-eval'; style-src 'self' 'unsafe-inline'; media-src *; img-src 'self' data: content:;">

        -->

    <meta name="format-detection" content="telephone=no">
    <meta name="msapplication-tap-highlight" content="no">
    <meta name="viewport" content="initial-scale=1, width=device-width, viewport-fit=cover">
    <meta name="color-scheme" content="light dark">
    <link rel="stylesheet" href="../css/index.css">
    <link rel="stylesheet" href="../css/w3.css">
    <title>Hello World</title>
</head>

<body>
<div class="w3-bar">
    <h1>Exceptions</h1>
    <p>&nbsp;</p>
</div>

<div class="w3-container">
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">C++ exceptions are a way to handle errors that occur during program execution. They allow you to transfer control from one part of a program to another, and to handle the error in a centralized way.</span></span></span></p>

        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">To use exceptions in C++, you use the following keywords:</span></span></span></p>
        
        <ul style="list-style-type:disc">
            <li><span style="font-size:11.0pt"><strong><span style="font-family:Arial"><span style="color:#1f1f1f">try:</span></span></strong></span><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f"> This keyword marks the beginning of a block of code that could potentially throw an exception.</span></span></span></li>
            <li><span style="font-size:11.0pt"><strong><span style="font-family:Arial"><span style="color:#1f1f1f">catch:</span></span></strong></span><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f"> This keyword marks the beginning of a block of code that will handle a specific type of exception.</span></span></span></li>
            <li><span style="font-size:11.0pt"><strong><span style="font-family:Arial"><span style="color:#1f1f1f">throw:</span></span></strong></span><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f"> This keyword is used to throw an exception.</span></span></span></li>
        </ul>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">The following code shows an example of how to use exceptions:</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1967d2">#include &lt;iostream&gt;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#8430ce">using namespace </span><span style="color:#1967d2">std</span><span style="color:#1f1f1f">;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#8430ce">int </span><span style="color:#996900">main</span><span style="color:#444746">()</span><span style="color:#1f1f1f"> {</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; <span style="color:#8430ce">try</span><span style="color:#1f1f1f"> {</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; <span style="color:#5f6368">// Code that could potentially throw an exception.</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; <span style="color:#8430ce">int</span><span style="color:#1f1f1f"> x = </span><span style="color:#b55908">10</span><span style="color:#1f1f1f"> / </span><span style="color:#b55908">0</span><span style="color:#1f1f1f">;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">&nbsp; } </span><span style="color:#8430ce">catch</span><span style="color:#1f1f1f"> (exception&amp; e) {</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; <span style="color:#5f6368">// Handle the exception.</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; <span style="color:#1967d2">cout</span><span style="color:#1f1f1f"> &lt;&lt; </span><span style="color:#188038">&quot;Error: &quot;</span><span style="color:#1f1f1f"> &lt;&lt; e.what() &lt;&lt; </span><span style="color:#1967d2">endl</span><span style="color:#1f1f1f">;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">&nbsp; }</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; <span style="color:#8430ce">return </span><span style="color:#b55908">0</span><span style="color:#1f1f1f">;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">}</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">Output:</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">Error: division by zero</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">In this example, the try block contains code that could potentially throw an exception, such as dividing by zero. If an exception is thrown, the catch block will be executed to handle the exception.</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">You can also throw custom exceptions. To do this, you create a class that inherits from the std::exception class. The following code shows an example of how to create a custom exception:</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#8430ce">class</span><span style="color:#996900"> MyException : </span><span style="color:#8430ce">public </span><span style="color:#1967d2">std</span><span style="color:#1f1f1f">::exception {</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#8430ce">public</span><span style="color:#1f1f1f">:</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">&nbsp; MyException(</span><span style="color:#8430ce">const </span><span style="color:#1967d2">string</span><span style="color:#1f1f1f">&amp; message) : message(message) {}</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; <span style="color:#8430ce">const char</span><span style="color:#444746">* </span><span style="color:#996900">what</span><span style="color:#444746">() </span><span style="color:#8430ce">const override </span><span style="color:#1f1f1f">{</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; <span style="color:#8430ce">return</span><span style="color:#1f1f1f"> message.c_str();</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">&nbsp; }</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#8430ce">private</span><span style="color:#1f1f1f">:</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; <span style="color:#1967d2">string</span><span style="color:#1f1f1f"> message;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">};</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">The following code shows an example of how to throw a custom exception:</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1967d2">#include &lt;iostream&gt;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#8430ce">using namespace </span><span style="color:#1967d2">std</span><span style="color:#1f1f1f">;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#8430ce">int </span><span style="color:#996900">main</span><span style="color:#444746">()</span><span style="color:#1f1f1f"> {</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; <span style="color:#8430ce">try</span><span style="color:#1f1f1f"> {</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; <span style="color:#8430ce">if</span><span style="color:#1f1f1f"> (</span><span style="color:#b55908">10</span><span style="color:#1f1f1f"> &lt; </span><span style="color:#b55908">5</span><span style="color:#1f1f1f">) {</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; &nbsp; <span style="color:#8430ce">throw</span><span style="color:#1f1f1f"> MyException(</span><span style="color:#188038">&quot;10 is not less than 5.&quot;</span><span style="color:#1f1f1f">);</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">&nbsp; &nbsp; }</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">&nbsp; } </span><span style="color:#8430ce">catch</span><span style="color:#1f1f1f"> (MyException&amp; e) {</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; <span style="color:#5f6368">// Handle the custom exception.</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; &nbsp; <span style="color:#1967d2">cout</span><span style="color:#1f1f1f"> &lt;&lt; </span><span style="color:#188038">&quot;Error: &quot;</span><span style="color:#1f1f1f"> &lt;&lt; e.what() &lt;&lt; </span><span style="color:#1967d2">endl</span><span style="color:#1f1f1f">;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">&nbsp; }</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier">&nbsp; <span style="color:#8430ce">return </span><span style="color:#b55908">0</span><span style="color:#1f1f1f">;</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">}</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">Output:</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Courier"><span style="color:#1f1f1f">Error: 10 is not less than 5.</span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Calibri">&nbsp;</span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f"><strong>Benefits of using exceptions</strong></span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">There are several benefits to using exceptions in C++, including:</span></span></span></p>
        
        <ul style="list-style-type:disc">
            <li><span style="font-size:11.0pt"><strong><span style="font-family:Arial"><span style="color:#1f1f1f">Error handling:</span></span></strong></span><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f"> Exceptions provide a centralized way to handle errors. This makes your code easier to read and maintain.</span></span></span></li>
            <li><span style="font-size:11.0pt"><strong><span style="font-family:Arial"><span style="color:#1f1f1f">Robustness:</span></span></strong></span><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f"> Exceptions can help you to write more robust code by handling errors gracefully.</span></span></span></li>
            <li><span style="font-size:11.0pt"><strong><span style="font-family:Arial"><span style="color:#1f1f1f">Readability:</span></span></strong></span><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f"> Exceptions can make your code more readable by making it clear how errors are handled.</span></span></span></li>
        </ul>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f"><strong>Conclusion</strong></span></span></span></p>
        
        <p><span style="font-size:11.0pt"><span style="font-family:Arial"><span style="color:#1f1f1f">Exceptions are a powerful feature of C++ that can be used to write more robust, reliable, and readable code. By understanding how to use exceptions, you can write better C++ code.</span></span></span></p>
        
        <p>&nbsp;</p>
        
    </div>

    <script src="../cordova.js"></script>
    <script src="../js/index.js"></script>
</body>

</html>